pkg/services/service.go

```diff
@@ -0,0 +1,232 @@
+	NewSubServices func(logger.Logger) []Service
+	// Start is an optional hook called after starting SubServices.
+	Start func(context.Context) error
+	// Close is an optional hook called before closing SubServices.
+	Close func() error //TODO examples
+}
+
+// NewEngine returns a new Service defined by Config, and an Engine for managing health, goroutines, and logging.
+//   - You *should* embed the Service, in order to inherit the methods.
+//   - You *should not* embed the Engine. Use an unexported field instead.
+func (c Config) NewEngine(lggr logger.Logger) (Service, *Engine) {
+	s := c.new(lggr)
+	return s, &s.eng
+}
+
+// NewService returns a new Service defined by Config.
+//   - You *should* embed the Service, in order to inherit the methods.
+func (c Config) NewService(lggr logger.Logger) Service {
+	return c.new(lggr)
+}
+
+func (c Config) new(lggr logger.Logger) *service {
+	lggr = logger.Sugared(logger.Named(lggr, c.Name))
+	s := &service{
+		cfg: c,
+		eng: Engine{
+			StopChan:      make(StopChan),
+			SugaredLogger: logger.Sugared(lggr),
+			conds:         make(map[string]error),
+		},
+	}
+	s.eng.healthEvent = s.StateMachine.SvcErrBuffer.Append
+	if c.NewSubServices != nil {
+		s.subs = c.NewSubServices(lggr)
+	}
+	return s
+}
+
+type service struct {
+	StateMachine
+	cfg  Config
+	eng  Engine
+	subs []Service
+}
+
+// Ready implements [HealthReporter.Ready] and overrides and extends [utils.StartStopOnce.Ready()] to include [Config.SubServices]
+// readiness as well.
+func (s *service) Ready() (err error) {
+	err = s.StateMachine.Ready()
+	for _, sub := range s.subs {
+		err = errors.Join(err, sub.Ready())
+	}
+	return
+}
+
+// Healthy overrides [StateMachine.Healthy] and extends it to include Engine errors as well.
+// Do not override this method in your service. Instead, report errors via the Engine.
+func (s *service) Healthy() (err error) {
+	err = s.StateMachine.Healthy()
+	if err == nil {
+		err = s.eng.healthy()
+	}
+	return
+}
+
+func (s *service) HealthReport() map[string]error {
+	m := map[string]error{s.Name(): s.Healthy()}
+	for _, sub := range s.subs {
+		CopyHealth(m, sub.HealthReport())
+	}
+	return m
+}
+
+func (s *service) Name() string { return s.eng.SugaredLogger.Name() }
+
+func (s *service) Start(ctx context.Context) error {
+	return s.StartOnce(s.cfg.Name, func() error {
+		s.eng.Info("Starting")
+		if len(s.subs) > 0 {
+			var ms MultiStart
+			s.eng.Infof("Starting %d sub-services", len(s.subs))
+			for _, sub := range s.subs {
+				if err := ms.Start(ctx, sub); err != nil {
+					s.eng.Errorw("Failed to start sub-service", "error", err)
+					return fmt.Errorf("failed to start sub-service of %s: %w", s.cfg.Name, err)
+				}
+			}
+		}
+		if s.cfg.Start != nil {
+			if err := s.cfg.Start(ctx); err != nil {
+				return fmt.Errorf("failed to start service %s: %w", s.cfg.Name, err)
+			}
+		}
+		s.eng.Info("Started")
+		return nil
+	})
+}
+
+func (s *service) Close() error {
```

Comment:
@DeividasK We could prioritize Stop while adapting Close easily here without spamming it in every implementation.
```suggestion
func (s *service) Close() error { return s.Stop() }
func (s *service) Stop() error {
```